test/e2e/garbage_collector.go

```diff
@@ -275,4 +329,123 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
+			rsList, err := rsClient.List(v1.ListOptions{})
+			if err != nil {
+				return false, fmt.Errorf("Failed to list rs: %v", err)
+			}
+			return len(rsList.Items) > 0, nil
+
+		})
+		if err == wait.ErrWaitTimeout {
+			err = fmt.Errorf("Failed to wait for the Deployment to create some ReplicaSet: %v", err)
+		}
+
+		By("delete the deployment")
+		deleteOptions := getNonOrphanOptions()
+		deleteOptions.Preconditions = v1.NewUIDPreconditions(string(createdDeployment.UID))
+		if err := deployClient.Delete(deployment.ObjectMeta.Name, deleteOptions); err != nil {
+			framework.Failf("failed to delete the deployment: %v", err)
+		}
+		By("wait for all rs to be garbage collected")
+		err = wait.PollImmediate(500*time.Millisecond, 1*time.Minute, func() (bool, error) {
+			return verifyRemainingDeploymentsAndReplicaSets(f, clientSet, deployment, 0, 0)
+		})
+		if err == wait.ErrWaitTimeout {
+			err = fmt.Errorf("Failed to wait for all rs to be garbage collected: %v", err)
+			remainingRSs, err := rsClient.List(v1.ListOptions{})
+			if err != nil {
+				framework.Failf("failed to list RSs post mortem: %v", err)
+			} else {
+				framework.Failf("remaining rs are: %#v", remainingRSs)
+			}
+
+		}
+
+		gatherMetrics(f)
+	})
+
+	It("should orphan RS created by deployment when deleteOptions.OrphanDependents is true", func() {
+		clientSet := f.ClientSet
+		deployClient := clientSet.Extensions().Deployments(f.Namespace.Name)
+		rsClient := clientSet.Extensions().ReplicaSets(f.Namespace.Name)
+		deploymentName := "simpletest.deployment"
+		deployment := newOwnerDeployment(f, deploymentName)
+		By("create the deployment")
+		createdDeployment, err := deployClient.Create(deployment)
+		if err != nil {
+			framework.Failf("Failed to create deployment: %v", err)
+		}
+		// wait for deployment to create some rs
+		By("Wait for the Deployment to create new ReplicaSet")
+		err = wait.PollImmediate(500*time.Millisecond, 1*time.Minute, func() (bool, error) {
+			rsList, err := rsClient.List(v1.ListOptions{})
+			if err != nil {
+				return false, fmt.Errorf("Failed to list rs: %v", err)
+			}
+			return len(rsList.Items) > 0, nil
+
+		})
+		if err == wait.ErrWaitTimeout {
+			err = fmt.Errorf("Failed to wait for the Deployment to create some ReplicaSet: %v", err)
+		}
+
+		By("delete the deployment")
+		deleteOptions := getOrphanOptions()
+		deleteOptions.Preconditions = v1.NewUIDPreconditions(string(createdDeployment.UID))
+		if err := deployClient.Delete(deployment.ObjectMeta.Name, deleteOptions); err != nil {
+			framework.Failf("failed to delete the deployment: %v", err)
+		}
+		By("wait for 2 Minute to see if the garbage collector mistakenly deletes the rs")
+		err = wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
+			ret, err := verifyRemainingDeploymentsAndReplicaSets(f, clientSet, deployment, 0, 1)
+			if err != nil {
+				return ret, err
+			}
+			if ret {
```

Comment:
Can you move this entire block out of the Poll()? Deployment shouldn't be deleted if the RS.ownerRefs are not cleaned up yet.